apps/arweave/src/ar_http_iface_server.erl

```diff
@@ -95,8 +94,181 @@ start_http_iface_listener(Config) ->
 	end,
 	ok.
 
+%%--------------------------------------------------------------------
+%% @doc stop the application and execute the shutdown procedure.
+%% @end
+%%--------------------------------------------------------------------
 stop() ->
-	cowboy:stop_listener(ar_http_iface_listener).
+	{ok, Config} = application:get_env(arweave, config),
+	Delay = Config#config.shutdown_tcp_connection_timeout,
+	ok = ranch:suspend_listener(ar_http_iface_listener),
+	Pids = terminate_connections(Delay),
+	terminate_listener(Pids, Delay).
+
+%%--------------------------------------------------------------------
+%% @doc ensure all connections have been closed before stopping the
+%%      cowboy listener.
+%% @end
+%%--------------------------------------------------------------------
+terminate_listener(Pids, Delay) ->
+	AfterDelay = Delay*3,
+	receive
+		{'DOWN', Ref, process, Pid, _} ->
+			Filter = fun({Pid, Ref}) -> false; (_) -> true end,
+			NewPids = lists:filter(Filter, Pids),
+			terminate_listener(NewPids, Delay);
+		_ -> terminate_listener(Pids, Delay)
```

Comment:
This could potentially loop, but the `after AfterDelay` limits the looping to `Delay*3`ms, right?

Reply:
oh wait, no it doesn't. 

Any risk to infinite loop here?